subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/PublishArtifactLocalArtifactMetadataCodec.kt

```diff
@@ -47,6 +49,49 @@ object PublishArtifactLocalArtifactMetadataCodec : Codec<PublishArtifactLocalArt
         val componentId = read() as ComponentIdentifier
         val ivyName = read() as IvyArtifactName
         val file = readFile()
-        return PublishArtifactLocalArtifactMetadata(componentId, DefaultPublishArtifact(ivyName.name, ivyName.extension, ivyName.type, ivyName.classifier, null, file))
+        return PublishArtifactLocalArtifactMetadata(componentId, ImmutablePublishArtifact(ivyName.name, ivyName.extension!!, ivyName.type, ivyName.classifier, file))
+    }
+}
+
+
+private
+class ImmutablePublishArtifact(
+    private val name: String,
+    private val extension: String,
+    private val type: String,
+    private val classifier: String?,
+    private val file: File,
+) : AbstractPublishArtifact() {
```

Comment:
❌ `AbstractPublishArtifact` has some mutable stuff in it and might diverge further. `ImmutablePublishArtifact` should  implement the interface directly.

And a data class would remove the need for equals and hashCode as well. What about something like the following?

```kotlin
private
data class ImmutablePublishArtifact(
    private val name: String,
    private val extension: String,
    private val type: String,
    private val classifier: String?,
    private val file: File,
) : PublishArtifactInternal {
    override fun getName() = name
    override fun getExtension() = extension
    override fun getType() = type
    override fun getClassifier() = classifier
    override fun getFile() = file
    override fun getDate(): Date? = null
    override fun shouldBePublished() = true
    override fun getBuildDependencies(): TaskDependency = TaskDependencyInternal.EMPTY
}
```

Reply:
Done, thanks!